lib/Common/Common/Jobs.h

```diff
@@ -503,6 +498,7 @@ namespace JsUtil
         virtual void AddManager(JobManager *const manager) override;
         virtual void RemoveManager(JobManager *const manager) override;
         template<class Fn> void ForEachManager(Fn fn); //This takes lock for criticalSection
+        template<class Fn> void ForEachJob(Fn fn); //This takes lock for criticalSection
```

Comment:
I see that you're following the pattern set by ForEachManager, but this makes me uncomfortable because the behavior of calling ForEachJob will depend on the compile-time type of reference the caller is holding, not the runtime type of the object being called. This is possible to use correctly, but also pretty easy to shoot one's foot with, in ways that will only manifest in subtle and impossible-to-reproduce race conditions. A possible solution would be to slightly break separation of concerns by having the base implementation of ForEachJob call a virtual private method that returns a unique_ptr<AutoCriticalSection>. Base implementation of that method would return nullptr and this class could override it to return a real lock.

Reply:
And today I learned we also have AutoOptionalCriticalSection for this kind of thing :)

---
In reply to: [185031871](https://github.com/Microsoft/ChakraCore/pull/5058#discussion_r185031871) [](ancestors = 185031871)

Reply:
agreed-- and thanks for finding AutoOptionalCriticalSection